### ISSUE TITLE
Interceptor presets are no longer accessible for import

### ISSUE THREAD

**Issue by cdimitroulas** [mswjs/interceptors#345]
See this comment: https://github.com/mswjs/interceptors/pull/313#issuecomment-1436419369

It seems that when we added bundling for ESM/CJS, we didn't consider some of the imports that are documented in the project's documentation that say that presets should be imported like so:
```ts
import browserInterceptors from '@mswjs/interceptors/lib/presets/browser'
```

**Comment by kettanaito** [comment]
Hi, @cdimitroulas. This makes sense, I think we forgot the "/presets" exports. Is this only the documentation problem though?

**Comment by cdimitroulas** [comment]
I haven't reviewed the documentation to see if other things are invalid, I was just opening this issue based on the comment I linked so that it doesn't get lost since the comment was on a closed PR.

**Comment by blaenk** [comment]
Yeah thank you @cdimitroulas I was about to create a tracking issue as well. Subscribed!

**Comment by kettanaito** [comment]
@blaenk, thanks for discovering this. Please, would you be interested in opening a pull request? This should be a matter of adding that path to those presets in "exports" in package.json. I can help you with the code review and getting this fix released.

**Comment by blaenk** [comment]
Apologies @kettanaito, this slipped, didn't mean to leave you hanging. I had responded to a similar request here https://github.com/mswjs/interceptors/pull/313#issuecomment-1437730466 and didn't see this one.

Thank you so much for working on it!

**Comment by kettanaito** [comment]
## Released: v0.22.5 🎉

This has been released in v0.22.5!

- 📄 [**Release notes**](https://github.com/mswjs/interceptors/releases/tag/v0.22.5)
- 📦 [npm package](https://www.npmjs.com/package/@mswjs/interceptors/v/0.22.5)

Make sure to always update to the latest version (`npm i @mswjs/interceptors@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Comment by kettanaito** [comment]
Hey, @blaenk, no worries! I think presents should be correctly exported now. Thanks for keeping an eye on these changes. I feel really good about the exports we've got now!

**Comment by blaenk** [comment]
Hey there, it's been a while and I just recently upgraded this package from the version I pinned from before this issue, and I noticed that unfortunately there is still an issue with the exports.

Running `tsc` directly also leads to this issue:

```
error TS2307: Cannot find module '@mswjs/interceptors/presets/browser' or its corresponding type declarations.

2 import browserInterceptors from "@mswjs/interceptors/presets/browser";
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I've spent a few hours fiddling with the thing, reading up on typescript documentation on `moduleResolution`, `module`, etc., and my vague hunch is increasingly that it's something to do with the conditional exports. I've run `tsc --traceResolution` to log out what exactly tsc is doing, and have tried this with all manner of combinations of configuration values, both in `tsconfig.ts` and while editing `@mswjs/interceptors`' own `package.json`.

I can't figure out exactly what is going wrong, but I have a vague feeling that there are two contributing factors:

1. It seems like the types that get applied up are the `node` ones, so then if I somehow force it to pick up the browser presets (e.g. by hand-editing this package's `package.json` to remove the `browser` conditional, which seems to be getting skipped otherwise), I get type errors seemingly because the types that the `node` types expect, while named the same, are not the same ones as those used in the 'browser' section of the library
2. For some reason, the `browser` condition in the `exports` is not triggering. I can't figure out how to get `tsc` to look inside it. I did see this potentially relevant part of the [node documentation](https://nodejs.org/api/packages.html#conditional-exports):

    > When using environment branches, always include a "default" condition where possible. Providing a "default" condition ensures that any unknown JS environments are able to use this universal implementation, which helps avoid these JS environments from having to pretend to be existing environments in order to support packages with conditional exports. For this reason, using "node" and "default" condition branches is usually preferable to using "node" and "browser" condition branches.

    I actually tried various combinations of this, and wasn't able to get it to work.

I think maybe the `exports` and use of conditionals there is too clever for the JS' ecosystem of tooling (even putting aside the dimensionality of ESM/CJS/etc.). I'm not sure what the solution is as I'm pretty new to this aspect of things and have spent many hours already trying to understand the problem.

I think it would be worth beginning by adding a sub-project to this repo which simply consists of the browser preset example, and hooking that up to github actions so we can ensure going forward that it keeps working.

**Comment by kettanaito** [comment]
@blaenk, I believe that's rather a documentation issue on the Node's side. I find the `exports` of Interceptors to be correct intention-wise. Certain presets mustn't be imported in certain environments. Restricting that on the imports level seems like a good idea. 

What ends up happening is people importing browser-specific presents in a Node.js module. Or, at least, in a module that TypeScript treats as as Node.js module. I believe this is also what's happening in your case. You should configure TypeScript to treat that module, and import resolution in it, as a browser module (if it's indeed intended for the browser, of course). If you share more information about what you're trying to do, I may help with more specific advice.

> I think maybe the exports and use of conditionals there is too clever for the JS' ecosystem of tooling 

I don't think so. The problem is that JavaScript developers aren't accustomed to differentiating between different environments while writing TypeScript. They should be, and I'd rather put my effort into explaining how to do that. 

I've adopted the new version of Interceptors across multiple frameworks and encountered no issues. Take a look at the [examples](https://github.com/mswjs/examples-new) to see if there are any extra steps necessary to make it work.

**Comment by blaenk** [comment]
> @blaenk, I believe that's rather a documentation issue on the Node's side. I find the exports of Interceptors to be correct intention-wise. Certain presets mustn't be imported in certain environments. Restricting that on the imports level seems like a good idea.

I completely empathize and agree. The intent seems correct. I understand the reason for wanting to use conditionals here.

> What ends up happening is people importing browser-specific presents in a Node.js module. Or, at least, in a module that TypeScript treats as as Node.js module. I believe this is also what's happening in your case.

I agree, that's the impression I was getting because the `--traceResolution` was indeed showing `node` paths.

> You should configure TypeScript to treat that module, and import resolution in it, as a browser module (if it's indeed intended for the browser, of course).

That's what I was getting at, or trying to at least. It is not clear to me what TypeScript configuration (`tsconfig.ts`) would have it looking under the `browser` condition within `package.json`'s `exports`. It doesn't help that it's a very tricky thing to search for, since any information, search results, documentation---is clobbered by the `export` keyword in ESM. That's not your problem of course, nor am I complaining, just sharing that I have put in the effort, traced log output, searched around, etc., and it's still not quite clear to me. The closest I've arrived at is `moduleResolution: bundler`, but that still didn't seem to help.

You are correct that I am targeting the browser. My situation is that I'm using webpack to bundle things, however, to be very clear, I quickly ruled out webpack when I ran into the same situation with `tsc --build`. I wonder if `browser` would only get looked at by a bundler like webpack, but then since it runs `tsc`, that part is what is failing because it would appear `tsc` is unaware of the `browser` condition

> The problem is that JavaScript developers aren't accustomed to differentiating between different environments while writing TypeScript. They should be, and I'd rather put my effort into explaining how to do that.

Agreed. I'd love to hear anything you're willing to share. I didn't come here lightly at the slightest sign of trouble, but only after really investing time to figure it out. By the way, when I mention the time I have spent trying to figure this out, I don't mean that at all to complain or to imply that it is the fault of this library, I'm grateful for the library. I mean it as a matter of fact, as I've been trying to get steeped in the whole ESM/CJS dichotomy and `exports` fields and dimensionality of tsc vs node vs bundlers, etc.

> I've adopted the new version of Interceptors across multiple frameworks and encountered no issues. Take a look at the [examples](https://github.com/mswjs/examples-new) to see if there are any extra steps necessary to make it work.

**Comment by blaenk** [comment]
> I've adopted the new version of Interceptors across multiple frameworks and encountered no issues. Take a look at the [examples](https://github.com/mswjs/examples-new) to see if there are any extra steps necessary to make it work.

Thank you so much I was not aware of that.

However, I did a quick search and saw that you are not using the browser presets? In other words there is no import like:

```ts
import browserInterceptors from '@mswjs/interceptors/presets/browser'
```

I should've been clearer: the problem I'm encountering appears to be most apparent when importing the browser presets.

I can 'circumvent' the problem (with a major caveat) by simply foregoing the browser preset (which is just an array) and importing the separate packages and constructing the array myself:

```ts
import { BatchInterceptor } from "@mswjs/interceptors";
import { XMLHttpRequestInterceptor } from "@mswjs/interceptors/XMLHttpRequest";
import { FetchInterceptor } from "@mswjs/interceptors/fetch";

const interceptor = new BatchInterceptor({
  name: "interceptor",
  interceptors: [new FetchInterceptor(), new XMLHttpRequestInterceptor()],
});
```

My suspicion though is that that is indeed doing what you suspect as well: that it is actually picking up the `node` types/packages. That said, it happens to actually work out seemingly fine in the browser, but of course I'd prefer to fix that.

So to put it plainly: I'd love to know if you'd be so kind to share what TypeScript configuration would be necessary so that when my project is built, it correctly looks into the `browser` conditions on the `exports`.

I have:

* `"type": "module",` in my `package.json`
* in my `tsconfig.ts` `compilerOptions` I have:

    ```json
    "moduleResolution": "bundler",
    "module": "ES2020",
    "target": "ES2020",
    "esModuleInterop": true
    ```

    Along with these, which I'm not sure how relevant they are:

    ````json
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": false,

**Comment by kettanaito** [comment]
Thanks for a great discussion around this! 

My plan is to release the next version of MSW and invest some resources into educating people around treating different environments (browser/node) while authoring tests, since that's the main area where such import issues happen. 

> It is not clear to me what TypeScript configuration (tsconfig.ts) would have it looking under the browser condition within package.json's exports.

Your confusion is valid. This setup often spreads outside of TS as well, since certain tools like Jest or Vitest may use code transformers and have their own module resolution on top of what you have in `tsconfig.json`. 

It doesn't help that some tools have unintuitive defaults that make you go in circles before you realize what's going on. For example, JSDOM forces _browser_ imports despite you writing a Node-oriented code. I'm opting-out of this behavior by using the `testEnvironmentOptions.customExportConditions` option in `jest.config.js`:

https://github.com/mswjs/msw/blob/faa71c42430176b8fb4aa073e2260a783437f3e2/test/jest.config.js#L13-L21

I'd have to research this topic better to have a write-up of the steps needed to solve it. I don't have all the info right now, unfortunately.

**Comment by kettanaito** [comment]
@blaenk, I think you're right. Those examples are using MSW in the browser, and in that environment, it doesn't rely on Interceptors at all, and uses a Service Worker instead 👍 Perhaps that's the reason why the error you're experiencing never surfaced in those examples. 

We can take the first action here in introducing a [module compatibility test suite](https://github.com/mswjs/msw/tree/faa71c42430176b8fb4aa073e2260a783437f3e2/test/modules). I wrote one for MSW and it has already prevented a number of broken releases in regards to ESM compatibility. We need to write this test suite for Interceptors as well to prevent module-related issues. 

You can take a stab at this if it sounds interesting to you! I will always support with code review along the way.

**Comment by blaenk** [comment]
Thanks for the call out about Jest. That's exactly the kind of thing I mean with respect to the various layers. I'm at least aware of the complexity there even if I don't fully grasp every corner of it, I am aware enough of the complexity to respect it.

This is why in my case, I'm not even at the point of involving Jest, jsdom, ts-jest, etc.---I'm purely running `tsc --build` alone, without any TypeScript compiler plugins or anything like that. It's as "vanilla" as it can get. I'm not going through a bundler like webpack/rollup/vite, etc.

**Comment by kettanaito** [comment]
It's perplexing to me that `tsc` would make any `exports`-related decisions, to be honest. Conditional exports are meant to be resolved by the environment, not the transpiler. A transpiler has no clue where this code will run. This is twice as confusing since TS doesn't have any designated module resolution options in terms of "hey, I want this subset of modules to use browser exports". `moduleResolution` is a different thing altogether.

**Comment by blaenk** [comment]
Thanks for that insight. That's the impression I was getting as well, that TypeScript wasn't really paying attention to those things nor had a way to express which to look at.

So if we consider what you just said, and then we look at the effective, distributed `package.json` for this package:

```json
  "main": "./lib/node/index.js",
  "module": "./lib/node/index.mjs",
  "types": "./lib/node/index.d.ts",
  "exports": {
    ".": {
      "types": "./lib/node/index.d.ts",
      "require": "./lib/node/index.js",
      "import": "./lib/node/index.mjs",
      "default": "./lib/node/index.js"
    },
    "./ClientRequest": {
      "browser": null,
      "types": "./lib/node/interceptors/ClientRequest/index.d.ts",
      "require": "./lib/node/interceptors/ClientRequest/index.js",
      "import": "./lib/node/interceptors/ClientRequest/index.mjs",
      "default": "./lib/node/interceptors/ClientRequest/index.js"
    },
    "./XMLHttpRequest": {
      "browser": {
        "types": "./lib/browser/interceptors/XMLHttpRequest/index.d.ts",
        "require": "./lib/browser/interceptors/XMLHttpRequest/index.js",
        "import": "./lib/browser/interceptors/XMLHttpRequest/index.mjs",
        "default": "./lib/browser/interceptors/XMLHttpRequest/index.js"
      },
      "types": "./lib/node/interceptors/XMLHttpRequest/index.d.ts",
      "require": "./lib/node/interceptors/XMLHttpRequest/index.js",
      "import": "./lib/node/interceptors/XMLHttpRequest/index.mjs",
      "default": "./lib/node/interceptors/XMLHttpRequest/index.js"
    },
    "./fetch": {
      "browser": {
        "types": "./lib/browser/interceptors/fetch/index.d.ts",
        "require": "./lib/browser/interceptors/fetch/index.js",
        "import": "./lib/browser/interceptors/fetch/index.mjs",
        "default": "./lib/browser/interceptors/fetch/index.js"
      },
      "types": "./lib/node/interceptors/fetch/index.d.ts",
      "require": "./lib/node/interceptors/fetch/index.js",
      "import": "./lib/node/interceptors/fetch/index.mjs",
      "default": "./lib/node/interceptors/fetch/index.js"
    },
    "./RemoteHttpInterceptor": {
      "browser": null,
      "types": "./lib/node/RemoteHttpInterceptor.d.ts",
      "require": "./lib/node/RemoteHttpInterceptor.js",
      "import": "./lib/node/RemoteHttpInterceptor.mjs",
      "default": "./lib/node/RemoteHttpInterceptor.js"
    },
    "./presets/node": {
      "browser": null,
      "types": "./lib/node/presets/node.d.ts",
      "require": "./lib/node/presets/node.js",
      "import": "./lib/node/presets/node.mjs",
      "default": "./lib/node/presets/node.js"
    },
    "./presets/browser": {
      "browser": {
        "types": "./lib/browser/presets/browser.d.ts",
        "require": "./lib/browser/presets/browser.js",
        "import": "./lib/browser/presets/browser.mjs",
        "default": "./lib/browser/presets/browser.js"
      }
    }
  },
```

Would it not therefore pick up one of the top-level `module`/`main` (I believe depending on whether my package is `type: module`) along with `types`? If that's the case, then doesn't it stand to reason that TypeScript would therefore try to use the _node_ types for everything? Zoom in on those three fields:

```json
  "main": "./lib/node/index.js",
  "module": "./lib/node/index.mjs",
  "types": "./lib/node/index.d.ts",
```

They are all node. So TypeScript would (I'm guessing) take these as the 'entry point' into things, and by then it's game over for any browser types being accessible, is my understanding.

So is the solution to this something like, don't compile with TypeScript alone, but instead with the bundler. However, the bundler will also just call out to TypeScript (`ts-loader` in my case with WebPack) which will attempt to type-check/etc. and so the situation would break down in the same manner (I'm only guessing. I can confirm though that bundling also fails).

**Comment by blaenk** [comment]
For what it's worth: I just manually edited the package's `package.json` (as in, within `node_modules`) to this, and building works now:

Basically, I removed any `node`-only export, and in cases where there were browser and node, i made browser the default (though not via `default`). Everything appears to work perfectly. `tsc --build` passes, and bundling does too.

I don't presume to understand the intricacies, but thought it might be a revealing data point for you, since you appear to be more understanding of these things. My uninformed hunch is that `tsc` is being led astray somehow or is simply oblivious to `browser` (if that even matters, I have no clue).

```json
  "main": "./lib/browser/index.js",
  "module": "./lib/browser/index.mjs",
  "types": "./lib/browser/index.d.ts",
  "exports": {
    ".": {
      "types": "./lib/browser/index.d.ts",
      "require": "./lib/browser/index.js",
      "import": "./lib/browser/index.mjs",
      "default": "./lib/browser/index.js"
    },
    "./XMLHttpRequest": {
      "types": "./lib/browser/interceptors/XMLHttpRequest/index.d.ts",
      "require": "./lib/browser/interceptors/XMLHttpRequest/index.js",
      "import": "./lib/browser/interceptors/XMLHttpRequest/index.mjs",
      "default": "./lib/browser/interceptors/XMLHttpRequest/index.js"
    },
    "./fetch": {
      "types": "./lib/browser/interceptors/fetch/index.d.ts",
      "require": "./lib/browser/interceptors/fetch/index.js",
      "import": "./lib/browser/interceptors/fetch/index.mjs",
      "default": "./lib/browser/interceptors/fetch/index.js"
    },
    "./presets/browser": {
      "types": "./lib/browser/presets/browser.d.ts",
      "require": "./lib/browser/presets/browser.js",
      "import": "./lib/browser/presets/browser.mjs",
      "default": "./lib/browser/presets/browser.js"
    }
  },
```

**Comment by kettanaito** [comment]
TypeScript always assumes Node module resolution, afaik. There's `moduleResolution: bundler`, which is recommended in some cases, but I don't know much about it to know what exactly it does. 

In general, modern bundlers will pick up the `exports` fields. `tsc` is a transpiler, so I have no idea why it'd even look into `exports`. Chances are, it doesn't. It just assumes Node and picks up `main`. 

> For what it's worth: I just manually edited the package's package.json (as in, within node_modules) to this, and building works now:

It works for your use case but it breaks imports in Node.js. Interceptors is a Node-focused library, so its main entry must always point to a Node.js build.

**Comment by blaenk** [comment]
> It works for your use case but it breaks imports in Node.js. Interceptors is a Node-focused library, so its main entry must always point to a Node.js build.

Oh I didn't mean to imply that this should be the solution at all. I understand that great care and effort has been invested into making this library cater to both environments, and I am grateful for it considering I am a consumer of the secondary use case, according to you.

I only presented that as a data point in case it might be revealing to you as to what may be going on in my case, if not specifically, then broadly.

In other words, I thought that by seeing that, it might immediately reveal to you what may be going on, given you seem more knowledgeable about these things.

What you're saying tracks along another hunch I had. That the `exports` are looked at by webpack, so does that mean that a project using the browser part of this library cannot build with `tsc` alone? By that I mean, for example, running it even if only for type-checking (of course would ultimately want a bundler in the end for actual use in a browser).

**Comment by blaenk** [comment]
If anyone's curious: I was playing around with it a bit and found that I can also resolve the issue by explicitly setting tsconfig's `paths`. I'm not sure if this is standard practice for doing this kind of thing (as in, not a hack), and whether it should be expected of users to know to do that (I genuinely don't know):

```json
    "baseUrl": ".",
    "paths": {
      "@mswjs/interceptors": ["node_modules/@mswjs/interceptors/lib/browser"],
      "@mswjs/interceptors/*": ["node_modules/@mswjs/interceptors/lib/browser/*"]
    }
```

**Comment by blaenk** [comment]
Just extra info for anyone else that runs into this: I was paranoid that webpack was actually still including the node files. I confirmed this by setting:

```js
    stats: {
      logging: 'verbose',
      modules: true,
    },
```

Then running a build with `--json=stats.json` and finding occurrences to the interceptors node paths.

So I realized that I needed to inform webpack of the tsconfig `paths` I configured to specifically look at the browser files, so that webpack can automatically creates aliases for them, and I did so using the package `tsconfig-paths-webpack-plugin` with this configuration:

```js
      plugins: [
        new TsconfigPathsPlugin({
          logLevel: "INFO",
          mainFields: ["browser", "module"],
          extensions: [".js", ".ts", ".tsx"]
        })
      ]
```

---

I would be surprised if every user knows to do this for arbitrarily-deep transitive package dependencies that require it. I feel like there probably is or should be some automatic way that I can configure my package to be targeting 'web' and for it to automatically use the browser files, both in `tsc` and in `webpack`.

It seems to me that since it's `tsc` which errors out primarily, the issue is probably with the fact that the top-level `types` refers to the node ones.

Some quick/light searching led me to https://github.com/Microsoft/TypeScript/issues/29128 which appears to be requesting a way to express certain types for a node package and other types for a browser package, e.g. with nested conditions like `exports` has or via a `types-browser` field or something like that.

The issue appears to have been closed by this comment https://github.com/Microsoft/TypeScript/issues/29128#issuecomment-449798551 which I will quote here:

> If your package has different types in node vs the browser.... have you considered that you really have two separate packages? We assume that a single package has a single API.

This gets back to my comment above where I wondered if this would all be simplified (both from a package maintainer perspective and from a user perspective) if two packages were published.

I humbly put that idea forward for @kettanaito and others to consider. I'll create a separate issue to track that.

**Comment by blaenk** [comment]
Sorry if this is considered spammy, but figured it may be interesting for anyone interested.

I learned about the `customConditions` setting in tsconfig. With it I can specify that it should consider `browser` true when looking at `exports` conditionals. This actually half worked in that I can correctly import the `@mswjs/interceptors/presets/browser` file now, but it overall fails because typescript is still using the node version of `index.d.ts`, so the type of `BatchInterceptor` is that of the `lib\node` version.

```json
"customConditions": ["browser"]
```

Like that typescript maintainer comment says, I wonder if it might be possible to unify the types somehow (so `browserInterceptors` might be named `commonInterceptors` or something) or indeed have separate packages. See again https://github.com/Microsoft/TypeScript/issues/29128

**Comment by blaenk** [comment]
Okay. adding that above condition to my `tsconfig.ts` (which I think is perfectly fine and probably preferable) + modifying this package's `package.json` to conditionally expose the browser types/files makes everything work! I'm curious what you think @kettanaito:

```json
    ".": {
      "browser": {
        "types": "./lib/browser/index.d.ts",
        "require": "./lib/browser/index.js",
        "import": "./lib/browser/index.mjs",
        "default": "./lib/browser/index.js"
      },
      "types": "./lib/node/index.d.ts",
      "require": "./lib/node/index.js",
      "import": "./lib/node/index.mjs",
      "default": "./lib/node/index.js"
    },
```

I will file a PR for discussion.

**Comment by blaenk** [comment]
See https://github.com/mswjs/interceptors/pull/421
